### PR TITLE
Add optional attendee feature

### PR DESCRIPTION
**Pull request by shulijones**
Add a feature to try to schedule meetings at times that optional attendees can make. Also add tests to confirm that this feature works as expected.

This passes all tests.

### CODE REVIEW

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 1]
```diff
@@ -124,6 +126,84 @@ public void everyAttendeeIsConsidered() {
...
+    // optional attendee who is busy during one of the original three
+    // meeting times. Only the other two should be returned.
+
+    Collection<Event> events = Arrays.asList(
+        new Event("Event 1", TimeRange.fromStartDuration(TIME_0800AM, DURATION_30_MINUTES),
+            Arrays.asList(PERSON_A)),
+        new Event("Event 2", TimeRange.fromStartDuration(TIME_0900AM, DURATION_30_MINUTES),
+            Arrays.asList(PERSON_B)),
+        new Event("Event 3", TimeRange.fromStartDuration(TIME_0830AM,
+        DURATION_30_MINUTES),
```
Comment (kyracq) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java]: nit: this could fit on the previous line

**Review by kyracq — APPROVED** [review]
LGTM! Just left one minor comment

**Review by anicr7 — APPROVED** [review]

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 2]
```diff
@@ -303,5 +408,27 @@ public void notEnoughRoom() {
 
     Assert.assertEquals(expected, actual);
   }
+
+  @Test
+  public void bothOptionalsTooBusy() {
+    // No mandatory attendees, and two optional attendees who are busy all day. // The result should be that there is no possible meeting time.
```
Comment (otobin) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java]: Nit: the two comments should be on separate lines

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 3]
```diff
@@ -23,12 +23,31 @@
...
+
+    Collection<TimeRange> mandatoryTimeOptions = queryMandatoryAttendees(events, mandatoryAttendees, longDuration);
+    Collection<TimeRange> optionalTimeOptions = queryMandatoryAttendees(events, allAttendees, longDuration);
 
+    if (!optionalTimeOptions.isEmpty() || mandatoryAttendees.isEmpty()) { 
+      // We are able to find times that work for both groups of attendees
+      return optionalTimeOptions;
+    }
+    return mandatoryTimeOptions;
+  }
```
Comment (otobin) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:41]: Love how short and compact this function is!

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 4]
```diff
@@ -124,6 +126,84 @@ public void everyAttendeeIsConsidered() {
...
+            Arrays.asList(PERSON_A)),
+        new Event("Event 2", TimeRange.fromStartDuration(TIME_0900AM, DURATION_30_MINUTES),
+            Arrays.asList(PERSON_B)),
+        new Event("Event 3", TimeRange.fromStartDuration(TIME_0830AM,
+        DURATION_30_MINUTES), 
+            Arrays.asList(PERSON_C)));
+
+    MeetingRequest request =
+        new MeetingRequest(Arrays.asList(PERSON_A, PERSON_B), DURATION_30_MINUTES);
+      request.addOptionalAttendee(PERSON_C);
```
Comment (otobin) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java]: nit: I think there's an extra space here

**Review by otobin — APPROVED** [review]
LGTM! A few really small comments

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java (hunk 1 above)
Comment (shulijones) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java]: Changed

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java (hunk 2 above)
Comment (shulijones) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java]: Fixed, thanks!

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java (hunk 4 above)
Comment (shulijones) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java]: Fixed

**Pull request merged**